src/dsl.jl

```diff
@@ -717,31 +703,38 @@ function read_equations_options(options, syms_declared, parameters_extracted; re
     # Loops through all equations, checks for lhs of the form `D(X) ~ ...`.
     # When this is the case, the variable X and differential D are extracted (for automatic declaration).
     # Also performs simple error checks.
-    vars_extracted = OrderedSet{Union{Symbol, Expr}}()
+    vs_inferred = OrderedSet{Union{Symbol, Expr}}()
     add_default_diff = false
     for eq in equations
         if (eq.head != :call) || (eq.args[1] != :~)
             error("Malformed equation: \"$eq\". Equation's left hand and right hand sides should be separated by a \"~\".")
         end
 
-        # If the default differential (`D`) is used, record that it should be decalred later on.
-        if (:D ∉ union(syms_declared, parameters_extracted)) && find_D_call(eq)
+        # If the default differential (`D`) is used, record that it should be declared later on.
+        if (:D ∉ syms_unavailable) && find_D_call(eq)
             requiredec && throw(UndeclaredSymbolicError(
                 "Unrecognized symbol D was used as a differential in an equation: \"$eq\". Since the @require_declaration flag is set, all differentials in equations must be explicitly declared using the @differentials option."))
             add_default_diff = true
-            push!(syms_declared, :D)
+            push!(syms_unavailable, :D)
         end
 
-        # Any undecalred symbolic variables encountered should be extracted as variables.
-        add_syms_from_expr!(vars_extracted, eq, syms_declared)
-        (!isempty(vars_extracted) && requiredec) && throw(UndeclaredSymbolicError(
-            "Unrecognized symbolic variables $(join(vars_extracted, ", ")) detected in equation expression: \"$(string(eq))\". Since the flag @require_declaration is declared, all symbolic variables must be explicitly declared with the @species, @variables, and @parameters options."))
+        # Any undeclared symbolic variables encountered should be extracted as variables.
+        add_syms_from_expr!(vs_inferred, eq, syms_unavailable)
+        (!isempty(vs_inferred) && requiredec) && throw(UndeclaredSymbolicError(
+            "Unrecognized symbolic variables $(join(vs_inferred, ", ")) detected in equation expression: \"$(string(eq))\". Since the flag @require_declaration is declared, all symbolic variables must be explicitly declared with the @species, @variables, and @parameters options."))
+    end
+
+    # If `D` differential is used, add it to differential expression and inferred differentials list.
+    diffs_inferred = Union{Symbol, Expr}[]
+    if add_default_diff && !any(diff_dec.args[1] == :D for diff_dec in diffsexpr.args)
+        diffs_inferred = [:D]
+        push!(diffsexpr.args, :(D = Differential($(tiv))))
     end
```

Comment:
Now, if we use a default differential, `diffsexpr` is updated directly in `read_equations_options!`. Made more sense as I wanted to handle the `@differnetials` option at the beginning (since this explicitly declares symbols). Previously this option was handled after `@equations`.